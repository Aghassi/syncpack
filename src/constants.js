export const DEFAULT_PACKAGES = './packages/*/package.json';
export const DEFAULT_SOURCE = './package.json';
