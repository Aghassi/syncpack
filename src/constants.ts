import { IManifestKey } from './typings';

export const FIX_MISMATCHES = {
  description: 'set dependencies used with different versions to the same version',
  name: 'fix-mismatches'
};
export const LIST = {
  description: 'list every dependency used in your packages',
  name: 'list'
};
export const LIST_MISMATCHES = {
  description: 'list every dependency used with different versions in your packages',
  name: 'list-mismatches'
};
export const DEFAULT_PATTERN = './packages/*/package.json';
export const DEPENDENCY_TYPES: IManifestKey[] = ['dependencies', 'devDependencies', 'peerDependencies'];
export const GREATER = 1;
export const LESSER = -1;
export const OPTION_PACKAGES = {
  description: `location of packages, defaults to "${DEFAULT_PATTERN}"`,
  spec: '-p, --packages <glob>'
};
export const SAME = 0;
export const SEMVER_ORDER = ['<', '<=', '', '~', '^', '>=', '>', '*'];
