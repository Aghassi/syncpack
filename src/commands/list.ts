import chalk from 'chalk';
import { SyncpackConfig } from '../constants';
import { getWrappers, SourceWrapper } from './lib/get-wrappers';
import { getDependencies } from './lib/installations/get-dependencies';
import { sortByName } from './lib/installations/sort-by-name';
import { log } from './lib/log';

type Options = Pick<SyncpackConfig, 'dev' | 'filter' | 'peer' | 'prod' | 'source'>;

export const list = (wrappers: SourceWrapper[], options: Options): void => {
  const iterator = getDependencies(wrappers, options);
  const packages = Array.from(iterator).filter(({ name }) => name.search(new RegExp(options.filter)) !== -1);

  packages.sort(sortByName).forEach(({ name, installations }) => {
    const versions = installations.map(({ version }) => version);
    const uniques = Array.from(new Set(versions));
    const hasMismatches = uniques.length > 1;
    const uniquesList = uniques.sort().join(', ');
    const message = hasMismatches
      ? chalk`{red âœ• ${name}} {dim.red ${uniquesList}}`
      : chalk`{dim -} {white ${name}} {dim ${uniquesList}}`;
    log(message);
  });
};

export const listFromDisk = (options: Options): void => {
  const wrappers = getWrappers(options);
  list(wrappers, options);
};
