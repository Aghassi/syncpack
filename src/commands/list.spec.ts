import * as mock from '../../test/mock';
import { DEFAULT_CONFIG } from '../constants';
import * as api from './list';

describe('list', () => {
  let list: typeof api.list;
  let log: jest.Mock;

  afterEach(() => {
    jest.restoreAllMocks();
  });

  beforeEach(() => {
    jest.mock('./lib/log', () => ({ log: jest.fn() }));
    list = require('./list').list;
    log = require('./lib/log').log;
  });

  it('outputs all dependencies', () => {
    const wrappers = [mock.wrapper('a', ['foo@0.1.0']), mock.wrapper('b', ['foo@0.2.0'])];
    list(wrappers, { ...DEFAULT_CONFIG, dev: false, peer: false, prod: true });
    expect(log.mock.calls).toMatchSnapshot();
  });
});
