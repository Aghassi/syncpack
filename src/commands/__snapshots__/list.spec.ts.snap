// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list outputs all dependencies 1`] = `
Array [
  Array [
    "[31mâœ• foo[39m [2m[31m0.1.0, 0.2.0[39m[22m",
  ],
]
`;
